tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,481 @@
+        return unary_element_wise_operation([](U v) {return std::sqrt(v); }); // lambda used because types cannot be deduced by simply specifying std::sqrt 
+    }
+
+    /* @brief Element-wise exponential
+     */
+    Tensor exp() const {
+        return unary_element_wise_operation([](U v) {return std::exp(v); }); // lambda used because types cannot be deduced by simply specifying std::exp
+    }
+
+ private:
+     void toDevice() const {
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+         CLCudaAPI::Queue queue = device->queue();
+         if (device_data_ && device_data_->GetSize() >= host_data_.size()) {
+             device_data_->Write(queue, host_data.size(), host_data_.data(), 0);
+         } else {
+             CLCudaAPI::Context ctx = device->context();
+             device_data_ = make_unique<CLCudaAPI::Buffer<U> >(
+                 ctx, queue, host_data_.begin(), host_data_.end());
+         }
+
+#endif
+         data_is_on_host_ = false;
+         data_dirty_ = false;
+     }
+
+     void fromDevice() const {
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+         assert(device_);
+         assert(device_data_);
+         device_data_->Read(device_->queue(), host_data_.size(), const_cast<U*>(host_data_.data())); // using const_cast<> to avoid making host_data_ entirely mutable
+#endif
+         data_is_on_host_ = true;
+         data_dirty_ = false;
+     }
+
+    // Initializes the shape vector
+    void resize(const cnn_size_t d0,
+                 const cnn_size_t d1,
+                 const cnn_size_t d2,
+                 const cnn_size_t d3) {
+	    shape_[0] = d0;
+	    shape_[1] = d1;
+	    shape_[2] = d2;
+	    shape_[3] = d3;
+        host_data_.resize(std::accumulate(std::begin(shape_), std::end(shape_), size_t(1), std::multiplies<size_t>()), U(0));
+    }
+
+    template<typename F> Tensor binary_element_wise_operation(const Tensor &src, F f) const
+    {
+        if (this->shape() != src.shape()) {
+            throw nn_error("Tensor must have same shape");
+        }
+
+        Tensor<U> res(src.shape());
+
+        const U* dst = res.mutable_host_data();
+        const U* src1 = host_data();
+        const U* src2 = src.host_data();
+
+        for_i(true, res.size(), [dst, src1, src2, &f](size_t i) {
+            dst[i] = f(src1[i], src2[i]);
+        });
+
+        return std::move(res);
+    }
+
+    template<typename F> Tensor unary_element_wise_operation(F f) const
+    {
+        if (this->shape() != src.shape()) {
+            throw nn_error("Tensor must have same shape");
+        }
+
+        Tensor<U> res(src.shape());
+
+        const U* dst = res.mutable_host_data();
+        const U* src1 = host_data();
+
+        for_i(true, res.size(), [dst, src1, &f](size_t i) {
+            dst[i] = f(src1[i]);
+        });
+
+        return std::move(res);
+    }
+
+    template<typename F> Tensor binary_scalar_operation(U scalar, F f) const
+    {
+        if (this->shape() != src.shape()) {
```

Comment:
Where is src variable in this `binary_scalar_operation` method?
Unlike binary_element_wise_operation method, this doesn't have src argument.